lib/types.nix

```diff
@@ -537,6 +537,14 @@ rec {
       modules = toList modules;
     };
 
+    # A module to be imported in some other part of the configuration.
+    deferredModule = mkOptionType {
+      name = "deferredModule";
+      description = "module";
+      check = t: isAttrs t || isFunction t;
```

Comment:
Shouldn't this also allow paths?

Reply:
And how about also allowing lists of modules, instead of just a single one

Reply:
Hmm, it bothers me a bit that the result of this type is a list, but the input is a single element. This is in contrast to essentially all other types whose input and output types are generally the same.

Also, if the input type is also a list (which I think also makes more sense, since it's how `imports`, the `modules` argument for `evalModules`/`submoduleWith` works), then it might make sense to use the `listOf` type to implement it.

Which then brings me to my general dislike of lists, the main problem being that it's not possible to remove entries once added. I guess in this case it's somewhat fine though, because I believe `disabledModules` can be used to essentially remove modules from the list, though finding out how to actually refer to modules in the list is pretty hard.

Reply:
This reminds me of [`Html` in Elm](https://discourse.elm-lang.org/t/why-is-html-a-single-dom-node/3966/2), which is also an inherently monoidal type that they're constantly putting in lists (the free monoid) for no good reason. Similar to their `[Html] -> Html` operation (`<div>` or `<span>`), our concat operation (`x: { imports = x; }`) is also leaky. It leaks through in error messages with `anon-`-something-something strings.

I've embraced the leakiness and I've work with the rules:
 1. keep the conversions to a minimum (they're leaky)
 2. keep the interface simple
     - we don't have a use case for defining lists of modules, and it's still covered by `mkMerge`
     - `submodule` doesn't support lists, so why should `deferredModule`?

It'd be a different store if lists of modules were _always_ considered to be valid modules. This would include the `submodule` argument, `submodule` definitions, `imports` _items_, `evalModules { modules }` items, etc.


> my general dislike of lists

I share the dislike, as they imply ordering, which is often not necessary or desirable in a declarative system. In this case though, they are only as unnecessary and undesirable as in the rest of the modules system - not something we can expect to solve here.



> `disabledModules`

The best solution to removals is not to include the things in the first place.
`disabledModules` should operate just fine on the parent module and on path-imported ones. Anonymous ones will be equally problematic as anonymous submodules.


> though finding out how to actually refer to modules in the list is pretty hard.

Do you mean list indexing? That's not something I'd think of, but I guess someone could try - and eventually fail horribly.

I guess this might be reason to make it slightly harder? Although `x.imports !! y` isn't much of an obstacle over `x !! y`., especially compared to writing the function that only removes one element. I don't think we have that in `lib`?

I suppose I'm not entirely opposed to returning a single module. Given my argumentation, do you still think I should do it? (or something else)

Reply:
> I suppose I'm not entirely opposed to returning a single module. Given my argumentation, do you still think I should do it? (or something else)

I haven't considered that, but I really do like that idea! By returning a single module we avoid lists in the type sense while keeping the input and output types the same, that's a :+1: :+1: from me! It also aligns with the type name of `types.deferredModule` instead of `types.deferredModule`**`s`**